### ISSUE TITLE
AuthComponent redirect to referer loop

### ISSUE THREAD

**Issue by doshu** [cakephp/cakephp#10635]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.4
* Platform and Target: classic lamp env

in the latest version of cake has been introduced the ability to redirect the logged in user to the latest requested page before the user has been redirected to the login page.
this feature is very helpful but can introduce a login loop. this is the scenario:
the target user open a link (someone sended it to him) to a page that he is not allowed to see.
the authcomponent redirect the user to the login page, setting the redirect query variable to the not allowed page. 
the user put his credentials and is redirected to the not allowed page because of the redirect variable.
the page is not accessible and the authcomponent redirect to him the login page again
the user enters in a loop.

if the user is not expert he don't understand that the problem is a not allowed inital link.

is there the possibility to disable this feature in some conditions, for example in the isAuthorized controller action?

**Comment by dereuromark** [comment]
The login redirect only happens on POST, thus there cannot be a login loop.
Once you are back on login you see the form again by default. And have to try to login again to get redirected.
How did you get that one?

**Comment by doshu** [comment]
the login loop is not intended as an http redirect loop.
the user after the login will be redirected to the page pointed by the
redirect query variable
if the page that the redirect variable point to is forbidden for the
current user, the user will be redirected to the login form again, with the
redirect variable setted to the non accessible page
this over and over again

the redirect query variable should not be setted if the authcomponent
redirect the user to the login form after open a not accessible url.


2017-05-09 19:43 GMT+02:00 Mark Sch. <notifications@github.com>:

**Comment by dereuromark** [comment]
What is your login action code?
The **redirectUrl()** method should only be invoked inside a POST - and a successful login POST for that matter.

**Comment by doshu** [comment]
yes, the redirectUrl method is invoked only once after the POST of the
login action.
the redirectUrl method, inside the POST, redirect the user to the url
pointed by the redirect variable.
if the isAuthorized method return false, the authcomponent redirect
automatically the user to the login page again, with the redirect variable
setted.


2017-05-09 20:08 GMT+02:00 Mark Sch. <notifications@github.com>:

**Comment by dereuromark** [comment]
Try to set the `unauthorizedRedirect` config to what you need.
https://github.com/cakephp/cakephp/blob/master/src/Controller/Component/AuthComponent.php#L141-L146
I think this ticket is closable, do you agree?

**Comment by markstory** [comment]
> the redirect query variable should not be setted if the authcomponent redirect the user to the login form after open a not accessible url

Wouldn't this mean removing the redirect entirely? Not being logged in creates a non-accessible url.

**Comment by doshu** [comment]
authentication and authorization should be treat in different ways. if the user is not already logged in the current implementation is ok, but after the authentication, if there is an authorization issue too, the user should be redirected to a safe page (maybe an homepage) or to the referer, but only if the referer is not the login page.
if the user is already logged in and try to open a non-accessible url, should be redirected directly to the referer or another safe url

**Comment by markstory** [comment]
How would AuthComponent determine if a URL is safe?

**Comment by dereuromark** [comment]
As outlined above you can already set unauthorizedRedirect key to whatever you need.
For the default use case the current defaults and how it works seems fine to me.

**Comment by doshu** [comment]
the unauthorized redirect behavior should be personalizable. i should set
the safe url with the unauthorizedRedirect key, but choose if use it in
combination with redirecting to the referer, and eventually the redirect
query param. The best thing in my opinion is if the authcomponent allow to
set the order of these three params (unauthorizedRedirect, referer,
redirect), so if the referer is outside the app, or is the login page, the
unauthorizedRedirect key will be used. only if the unauthorizedRedirect is
not set, the redirect query param should be used. if the redirect query
params not exists too, the user will be redirected to the loginUrl

in this way we can avoid a login loop caused by the high priority that the
redirect query param has now.

2017-05-10 23:52 GMT+02:00 Mark Sch. <notifications@github.com>:

**Comment by dereuromark** [comment]
Feel free to open a PR here with your suggested changes for further discussion.

**Comment by saeideng** [comment]
similar to my old issue   https://github.com/cakephp/cakephp/issues/11067

**Comment by dereuromark** [comment]
@doshu Are you able to open a PR with your suggested changes? That will help a lot to get a clearer picture.
